### DISCUSSION TITLE
Resolution limit for vertical video.

### DISCUSSION THREAD

**Discussion by odkrys** [ReVanced/revanced-discussions#929] [Question] [answered]
https://www.youtube.com/watch?v=y6oMutwJQCw

Can I get vertical video to 2160p on a device with a resolution of FHD?

My two tablets with FHD resolution only show up to 1080p for this vertical video.
On the other hand, other tablet and smartphone with QHD resolution go up to 2160p.
All horizontal videos can be selected up to 4K, but in the case of vertical videos, it seems that there is a limit to the quality that can be selected depending on the device resolution.

https://imgur.com/4IL9k5l

**Comment by oSumAtrIX** [comment]
https://github.com/revanced/revanced-patches/issues/1329

**Reply by odkrys** [reply]
I know YouTube's video ID is confusing between horizontal and vertical videos.
This is a very old bug that has been going on for years and has been improved in recent youtube versions since shorts were added.
(For example, when I press 2160p on old versions of youtube vanced or old original youtube, 1080p played, and when I press 1440p, 720p played)

This seems to be another issue where the device is not loading 1440p, 2160p vertical video data from the server.

**Reply by odkrys** [reply]
@oSumAtrIX Now I know what the problem is.
The YouTube app reads the machine's hardware acceleration information and requests the video list to the server.
If the acceleration information is created based on the horizontal, the vertical video will only load at a lower resolution. 
For example:
```
        <MediaCodec name="OMX.qcom.video.decoder.vp9" type="video/x-vnd.on2.vp9" >
            <Quirk name="requires-allocate-on-input-ports" />
            <Quirk name="requires-allocate-on-output-ports" />
            <Limit name="size" min="64x64" max="3840x2160" />
            <Limit name="alignment" value="2x2" />
            <Limit name="block-size" value="16x16" />
            <Limit name="blocks-per-second" min="1" max="972000" />
            <Limit name="bitrate" range="1-100000000" />
            <Feature name="adaptive-playback" />
            <Limit name="concurrent-instances" max="16" />
        </MediaCodec>
```
max="3840x2160" value is set and the request to youtube is
```
YT.qoe  : Pinging https://s.youtube.com/api/stats/qoe?cl=513398463&ei=70oIZP_DHMva2roP6--EqAY&event=streamingstats&feature=search&ns=yt&plid=AAX2X4ltRt_mcMnz&cpn=GgaW2UhR6-8pyjVd&cplatform=tablet&cbr=com.google.android.youtube&cff=LARGE_FORM_FACTOR&c=android&cmodel=MI%20PAD%204&cos=Android&csdk=27&cbrver=18.08.38&cver=18.08.38&cosver=8.1.0&cbrand=Xiaomi&vfs=0.693:244:243::i,0.883:244:244:244:r,0.987:244:247:244:r,6.409:244::244:M&df=0.883:0,0.987:0,8.766:0&conn=8.787:3&cplayer=ANDROID_EXOPLAYER_V2&bh=0.987:52.21,8.783:12.63&cache_bytes=0.863:2229902,8.783:11724016&cmt=8.783:7.060&afs=0.693:251::i&is_offline=0.520:0&sur=0.518:3,8.262:1&view=0.693:1200:675,0.883:423:754,0.987:687:1223&bat=8.798:0.450:1&rate=0.532:1.0&vps=0.527:B,0.863:PL,8.783:SU&bwm=8.783:6176087:0.501&user_intent=0.441&ctmp=pcmp:dh,vprng:minh.64;maxh.2160;minw.64;maxw.3840,soc:qcom;sdm660,mem:tot.40;max.512;free.5;heap.78;hpall.66;hpfree.11,mlat:rt.0.529;loadVideo.86296,llv:87,pmqs:dt.1153;lmq.2160;dir.1,cir:rt.0.689;reused.false;reason.CODEC_INIT_REASON_FIRST_PLAYBACK,vprng:minh.64;maxh.2160;minw.64;maxw.3840&seq=2&aur=0.866:5&docid=(scrubbed)&referrer=(scrubbed) 
```
```
minh.64;maxh.2160;minw.64;maxw.3840,soc:qcom;sdm660
```
<< height is limited to 2160 so only load up to 1080p. (1080x1920)

I changed the device's value to 3840x3840 and the youtube app's request also changed.

```
YT.qoe  : Pinging https://s.youtube.com/api/stats/qoe?adcontext=CAESEwiDwM7G_cv9AhXgYQ8CHYw6DCcgBCgBMAE&adformat=15_2_1&cl=513398463&ei=I1AIZO2IM_qT1d8PuMmd0AE&event=streamingstats&ns=yt&plid=AAX2X9jXFsxiB0qF&cpn=Uj-7_iRy3aFu_SPL&cplatform=tablet&cbr=com.google.android.youtube&cff=LARGE_FORM_FACTOR&c=android&cmodel=MI%20PAD%204&cos=Android&csdk=27&cbrver=18.08.38&cver=18.08.38&cosver=8.1.0&cbrand=Xiaomi&conn=0.009:3&vfs=0.385:247:247::i&df=6.391:0&cplayer=ANDROID_EXOPLAYER_V2&bh=6.394:0.00&cmt=6.394:6.060&error=0.011:onesie.mismatch::0.000:c.cpn_mismatch.bDQl7LdUtl443qNm.Uj-7_iRy3aFu_SPL&afs=0.385:251::i&is_offline=0.031:0&sur=0.029:3&view=0.385:1200:675&bat=0.007:0.460:1,6.399:0.460:1&rate=0.045:1.0&vps=0.038:B,0.426:PL,6.394:EN&user_intent=0.006&ctmp=pcmp:dh,vprng:minh.64;maxh.3840;minw.64;maxw.3840,soc:qcom;sdm660,mem:tot.45;max.512;free.6;heap.96;hpall.73;hpfree.22,mlat:rt.0.041;loadVideo.32486,llv:33,pmqs:dt.1153;lmq.1080;dir.1,cir:rt.0.101;reused.true;mode.CODEC_REUSE_MODE_REUSE_WITH_RECONFIGURATION&seq=2&aur=0.428:5&docid=(scrubbed) 
```
```
minh.64;maxh.3840;minw.64;maxw.3840,soc:qcom;sdm660
```
Now I can watch up to 2160p portrait video.
Could this change be done in revanced?

**Reply by oSumAtrIX** [reply]
I am not sure how it is useful to play higher resolutions on a device that does not support higher resolutions. Unless you are working on high-quality screens, there's little need for high definition. A patch is certainly possible if you find where the request is sent off.

**Reply by odkrys** [reply]
As you know, YouTube offers higher bitrates at higher resolutions.
This is fully felt on a tablet with a large screen of 10~11 inches or more.
And the biggest beneficiary of this is entry-level APs that only support hardware acceleration up to 1080p.
For example snapdragon 662, 680, unisoc T612, helio g99 etc.

Low-end tablets using these chipsets only load 1080p horizontal videos and 480p (480x854) vertical videos from the YouTube app because maxh is 1080.

Imagine viewing a 480x854 vertical video on an 11-inch FHD tablet. It's terrible. And this is actually happening.

Unfortunately, I'm just an end user, so I don't know which part to debug for the development of Revanced. I'll try it, but I will call this "feature-request."

**Reply by odkrys** [reply]
```
root@instance-20220814-1527:~/test/youtube_18.08.39# grep -r "maxh" ./*
./smali_classes5/zgl.smali:    const-string v0, ";maxh."
root@instance-20220814-1527:~/test/youtube_18.08.39# grep -r "maxw" ./*
./smali_classes5/zgl.smali:    const-string v0, ";maxw."
```
Could this be a starting point?
I'm not sure how Revanced's development is going.

**Reply by oSumAtrIX** [reply]
Yeah, this could be a starting point, please open an issue with this information in the revanced-patches repository.

**Comment by odkrys** [comment, marked as answer]
https://github.com/ReVanced/revanced-patches/issues/1278
